trulens_eval/trulens_eval/tru_llama.py

```diff
@@ -179,11 +265,326 @@ def query_with_record(self, str_or_query_bundle) -> Tuple[Response, Record]:
+
+    # Compatible with llama_index.chat_engine.types.BaseChatEngine.chat .
+    def chat_with_record(self, message: str,
+                         **kwargs) -> Tuple[AgentChatResponse, Record]:
+        assert isinstance(
+            self.app, llama_index.chat_engine.types.BaseChatEngine
+        )
+
+        # Wrapped calls will look this up by traversing the call stack. This
+        # should work with threads.
+        record: Sequence[RecordAppCall] = []
+
+        ret = None
+        error = None
+
+        start_time = None
+        end_time = None
+
+        cost = Cost()
+
+        try:
+            start_time = datetime.now()
+
+            ret, cost = Endpoint.track_all_costs_tally(
+                lambda: self.app.chat(message, **kwargs)
+            )
+
+            end_time = datetime.now()
+
+        except BaseException as e:
+            end_time = datetime.now()
+            error = e
+            logger.error(f"Engine raised an exception: {e}")
+            logger.error(traceback.format_exc())
+
+        assert len(record) > 0, "No information recorded in call."
+
+        ret_record_args = dict()
+
+        # TODO: generalize
+        ret_record_args['main_input'] = message
+
+        assert isinstance(ret, AgentChatResponse)
+
+        ret_record_args['main_output'] = ret.response
+
+        ret_record = self._post_record(
+            ret_record_args, error, cost, start_time, end_time, record
+        )
+
+        return ret, ret_record
+
+
+    # Compatible with llama_index.chat_engine.types.BaseChatEngine.achat .
+    async def achat_with_record(
+        self, message: str, **kwargs
+    ) -> Tuple[AgentChatResponse, Record]:
+        assert isinstance(
+            self.app, llama_index.chat_engine.types.BaseChatEngine
+        )
+
+        # Wrapped calls will look this up by traversing the call stack. This
+        # should work with threads.
+        record: Sequence[RecordAppCall] = []
+
+        ret = None
+        error = None
+
+        start_time = None
+        end_time = None
+
+        cost = Cost()
+
+        try:
+            start_time = datetime.now()
+
+            ret, cost = await Endpoint.atrack_all_costs_tally(
+                lambda: self.app.achat(message, **kwargs)
+            )
+
+            end_time = datetime.now()
+
+        except BaseException as e:
+            end_time = datetime.now()
+            error = e
+            logger.error(f"Engine raised an exception: {e}")
+            logger.error(traceback.format_exc())
+
+        assert len(record) > 0, "No information recorded in call."
+
+        ret_record_args = dict()
+
+        # TODO: generalize
+        ret_record_args['main_input'] = message
+
+        assert isinstance(ret, AgentChatResponse)
+
+        ret_record_args['main_output'] = ret.response
+
         ret_record = self._post_record(
             ret_record_args, error, cost, start_time, end_time, record
         )
 
         return ret, ret_record
```

Comment:
how much of this flow can we start modularizing?

Reply:
A lot. There is a draft PR getting into this.

Reply:
https://github.com/truera/trulens/pull/344